Fix performance issue with alloptions filter

jom opened the pull request:
This fixes a pretty serious performance issue I introduced years ago that filters every options call on `alloptions`. I don't think these fallback options are needed. These legacy options go back to to [pre-1.0.9](https://github.com/Automattic/sensei/blob/version/1.0.9/classes/class-woothemes-sensei-settings.php). Removing this would only affect sites who haven't saved their Sensei settings in the last 10 years.  I think the `course_completed_page_id` was a copy-and-paste error from when that feature was added. 

## Proposed Changes
* I've removed the legacy fallback for these options and the `alloptions` filter.
* Use `wp_load_alloptions` strategy on `sensei-install-version` which was querying on every page load if not set (Sensei was installed before this was added)

Note: `sensei-data-port-jobs` is still queried, but that is more complicated because we don't autoload it. Filed https://github.com/Automattic/sensei/issues/6760 for that one.

## Deprecations
- `Sensei_Settings::no_special_query_for_legacy_options`

## Testing Instructions
1. Activate query monitor.
2. Remove the `sensei-install-version`
3. Ensure there isn't a `{prefix}options` query for `woothemes-sensei_user_dashboard_page_id`, `woothemes-sensei_courses_page_id`, or `sensei-install-version` when visiting Sensei and/or non-Sensei screens in WP Admin.

## Pre-Merge Checklist
<!-- Complete applicable items on this checklist **before** merging. Items that are not applicable can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed. -->
- [x] PR title and description contain sufficient detail and accurately describe the changes
- [x] Acceptance criteria is met
- [ ] Decisions are publicly documented
- [x] Adheres to coding standards ([PHP](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/), [JavaScript](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/), [CSS](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/), [HTML](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/html/))
- [ ] All strings are translatable (without concatenation, handles plurals)
- [ ] Follows our naming conventions (P6rkRX-4oA-p2)
- [ ] Hooks (p6rkRX-1uS-p2) and functions are documented
- [ ] New UIs are responsive and use a [mobile-first approach](https://zellwk.com/blog/how-to-write-mobile-first-css/)
- [ ] New UIs match the designs
- [ ] Different user privileges (admin, teacher, subscriber) are tested as appropriate
- [ ] Code is tested on the minimum supported PHP and WordPress versions
- [ ] User interface changes have been tested on the latest versions of Chrome, Firefox and Safari
- [ ] "Needs Documentation" label is added if this change requires updates to documentation
- [ ] Known issues are created as new GitHub issues

jom committed:
Fix performance issue with alloptions filter

jom committed:
Add changelog

renatho commented:
The hook didn't run a lot of times as expected, but I the following point didn't work for me:

> Ensure there isn't a {prefix}options query for woothemes-sensei_user_dashboard_page_id or woothemes-sensei_courses_page_id when visiting Sensei and/or non-Sensei screens in WP Admin.

<img width="1482" alt="Screenshot 2023-04-04 at 15 35 23" src="https://user-images.githubusercontent.com/876340/229887252-1c066d01-15e5-48d7-8121-8ad7d5a22b40.png">

And, honestly, I didn't understand exactly the purpose of that solution. The only thing I could understand from that, is to avoid a query for that options, but I just found 1 `get_option` for each in the whole code.

fjorgemota commented:
I tested this PR, and I couldn't get the hook to be triggered...in the web server environment. I was able to make the code run when running on WP CLI, especially with commands like `wp cache flush`, but not on a webpage.

On the webpage, I found out that the query that @renatho mentioned indeed runs.

With that said, couldn't we maybe use the [`default_option_$option` hook](https://developer.wordpress.org/reference/hooks/default_option_option/) instead? I think it is made exactly for this use-case.. :thinking:

fjorgemota commented:
Ha, I figured out the mystery: By the time the `add_filter` runs, the `pre_cache_alloptions` hook was already triggered, as you can confirm by adding `var_dump(did_filter('pre_cache_alloptions'));` just after the `add_filter`.

So basically that filter never runs, and because of that the SQL query is run anyway. I'm testing on WP 6.1, by the way.

By using `alloptions`, the default values are indeed added, but at the cost of running the filter a LOT of times. I'm not sure exactly what's worse in this case, to be honest, given that WP itself has a cache for items that don't exist and will return the default value promptly if that's the case (which is filtered by the filter mentioned above, so forget about that "solution"), so maybe adding this filter has a bigger runtime cost than adding nothing at all.

If we REALLY want to avoid these queries, one option we could consider is replacing code to add the filter with adding the default values to the `options` cache group, like this:

```php
wp_cache_add( 'woothemes-sensei_user_dashboard_page_id', 0, 'options' );
wp_cache_add( 'woothemes-sensei_courses_page_id', 0, 'options' );
```

By doing this, WordPress will avoid the SQL query because it will look at the cache first before querying the database, but after checking the autoloaded options. So basically:

1. If we have these options auto-loaded by `wp_load_alloptions()`, everything is beautiful and WordPress will just return the auto-loaded value;
2. If we don't have these options auto-loaded by `wp_load_alloptions();`, WordPress will check the cache entry and, if it exists, return it;

In the worst case scenario (e.g. if WordPress changes its cache group or logic for caching), the SQL query will be made anyway. But, for now, it seems like a good idea.

WDYT?

jom commented:
@renatho @fjorgemota Thanks for your testing. I'm not sure why I wasn't seeing those queries locally. 

@fjorgemota I'll take a look at your suggestion. The other thing I was considering is to load `wp_load_alloptions()` when registering our settings and looking directly there for our legacy fallback option defaults.

jom committed:
Switch to new strategy for fetching legacy options

jom committed:
Remove fallback to legacy options

jom committed:
Update changelog

jom committed:
Add method back and deprecate it

jom commented:
@renatho @fjorgemota I switched my strategy in 159d76c9e142c45d31c428a48ee25c7b750b9db6, but when I was investigating to confirm these options were added with `autoload` enabled, I discovered this fallback has been happening since our earliest tagged version (1.0.9; [user_dashboard_page_id](https://github.com/Automattic/sensei/blob/version/1.0.9/classes/class-woothemes-sensei-settings.php#L144); [course_page_id](https://github.com/Automattic/sensei/blob/version/1.0.9/classes/class-woothemes-sensei-settings.php#L134)). I don't think we need to do all this weird stuff to support users who haven't saved their Sensei settings in 10 years. 😃 

I noticed we were querying for `sensei-install-version` so I switched to the `wp_load_alloptions` strategy in https://github.com/Automattic/sensei/pull/6753/commits/aecd4c0f5e3b3797f03998807a7a854e1d16a2a7

jom committed:
Fix PHPCS issue

jom committed:
Use alloptions strategy on sensei-install-version

jom committed:
Fix equals sign alignment

fjorgemota reviewed (approved):
I don't see a ton of improvement between this PR and what's on `trunk` right now. At least not on the CPU I'm using, the story might be different on a weaker CPU..

BUT, I'm also the one who asked you to change `str_contains` to `str_starts_with` just for the sake of optimization...

So, I'm approving this. It looks good, it works well, and it doesn't seem to break anything, which is great.

Thanks!

jom commented:
You'll only notice it in XDebug or Code Profiler. My last Code Profiler on SenseiLMS Staging had this alloptions callback taking up nearly ~0.07s of execution time, which is significant considering what it does. It was in the top 5 most time taking functions. Tick-based profilers like this often warn that the autoloader profiles aren't especially useful, so this was the most time-consuming method when you take out the autoloaders. 

<img width="1542" alt="Screenshot 2023-04-06 at 9 25 40 am" src="https://user-images.githubusercontent.com/68693/230319457-7bb26f27-4914-40c7-b5f6-22d60f18f002.png">

fjorgemota commented:
> You'll only notice it in XDebug or Code Profiler. My last Code Profiler on SenseiLMS Staging had this alloptions callback taking up nearly ~0.07s of execution time, which is significant considering what it does. It was in the top 5 most time taking functions. Tick-based profilers like this often warn that the autoloader profiles aren't especially useful, so this was the most time-consuming method when you take out the autoloaders.

Ouch, interesting. Yeah, I didn't test using a code profiler, my fault right there.

Thanks for the extra information! :D
